frame/transaction-storage/src/lib.rs

```diff
@@ -0,0 +1,364 @@
+		/// # <weight>
+		/// - n*log(n) of data size, as all data is pushed to an in-memory trie.
+		/// Additionally contains a DB write.
+		/// # </weight>
+		#[pallet::weight(T::WeightInfo::store(data.len() as u32))]
+		pub(super) fn store(
+			origin: OriginFor<T>,
+			data: Vec<u8>,
+		) -> DispatchResultWithPostInfo {
+			ensure!(data.len() > 0, Error::<T>::EmptyTransaction);
+			Self::apply_fee(origin, data.len() as u32)?;
+
+			// Chunk data and compute storage root
+			let chunks = data.chunks(CHUNK_SIZE).map(|c| c.to_vec()).collect();
+			let root = sp_io::trie::blake2_256_ordered_root(chunks);
+
+			let content_hash = sp_io::hashing::blake2_256(&data);
+			let extrinsic_index = <frame_system::Pallet<T>>::extrinsic_index().unwrap();
+			sp_io::transaction_index::index(extrinsic_index, data.len() as u32, content_hash);
+
+			let block_num = <frame_system::Pallet<T>>::block_number();
+			let counter = <Counter<T>>::mutate(|c| { *c += 1; *c - 1 });
+			<TransactionRoots<T>>::insert(block_num, counter, TransactionInfo {
+				chunk_root: root,
+				size: data.len() as u32,
+				content_hash: content_hash.into(),
+			});
+			Self::deposit_event(Event::Stored(counter));
+			Ok(().into())
+		}
+
+		/// Renew previously stored data. Parameters are the block number that contains
+		/// previous `store` or `renew` call and transaction index within that block.
+		/// Transaction index is emitted in the `Stored` or `Renewed` event.
+		/// Applies same fees as `store`.
+		/// # <weight>
+		/// - Constant.
+		/// # </weight>
+		#[pallet::weight(T::WeightInfo::renew())]
+		pub(super) fn renew(
+			origin: OriginFor<T>,
+			block: T::BlockNumber,
+			index: u32,
+		) -> DispatchResultWithPostInfo {
+			let info = <TransactionRoots<T>>::get(block, index).ok_or(Error::<T>::RenewedNotFound)?;
+			Self::apply_fee(origin, info.size)?;
+
+			let extrinsic_index = <frame_system::Pallet<T>>::extrinsic_index().unwrap();
+			sp_io::transaction_index::renew(extrinsic_index, info.content_hash.into());
+
+			let counter = <Counter<T>>::mutate(|c| { *c += 1; *c - 1 });
+			let block_num = <frame_system::Pallet<T>>::block_number();
+			<TransactionRoots<T>>::insert(block_num, counter, info);
+			Self::deposit_event(Event::Renewed(counter));
+			Ok(().into())
+		}
+
+		/// Check storage proof for block number `block_number() - StoragePeriod`.
+		/// If such block does not exist the proof is expected to be `None`.
+		/// # <weight>
+		/// - Linear w.r.t the number of indexed transactions in the proved block for random probing.
+		/// There's a DB read for each transaction.
+		/// Here we assume a maximum of 100 probed transactions.
+		/// # </weight>
+		#[pallet::weight((T::WeightInfo::check_proof_max(), DispatchClass::Mandatory))]
+		pub(super) fn check_proof(
+			origin: OriginFor<T>,
+			proof: Option<TransactionStorageProof>,
+		) -> DispatchResultWithPostInfo {
+			ensure_none(origin)?;
```

Comment:
What stops other people from calling this function?

Reply:
You mean users making a transaction with this call?
It's an inherent call. Doesn't `ensure_none` effectively check that? I've used the same setup as e.g. `timestamp::set`

Reply:
I think the difference is that `timestamp::set` has a check that timestamp was already set as an early exit, and part of the inherent pre-verification.

I think you can submit an unsigned extrinsic and it will pass the `ensure_none` check, but i dont exactly remember. Would be good to check again, because this function seems super heavy, and nothing is preventing it from running twice (or more) in a block afaik.